---
title: "Analysis of microbiome in anaerobic gut-on-chip"
author: "Diogo M. Camacho, Ph.D. w/ Jalili et al."
output: 
  html_document:
    toc: TRUE
    float_toc: TRUE
    code_folding: hide
    theme: readable
    df_print: paged
---


```{r setup, include=FALSE}
# general options
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)

source("./R/functions_libraries.R")
```


# Overview
Here we will recapitulate the analyses done for the manuscript.  These are the analyses on the 16S sequencing data on samples collected from gut-on-chip in different O2 concentrations (please revert to publication for details on experimental setup.)  All code is available on [Github]("http://github.com/diogocamacho/jalili_microbiome").  This report was prepared using `rmarkdown` and knitted into HTML using `knitr`.

### Data 
Data was processed with `QIIME` and loaded into R using the `phyloseq` package.  Data was subsequently transform into the `tidy` format using functions in the `tidyverse` package.

```{r data_prep}
load("data/unfiltered_data_2018-07-02.RData")
load("data/tidy_filtered_2018-07-02.RData")

OTU <- untidy_unfil[[1]]
full_OTU <- untidy_unfil[[2]]
HMPv35 <- untidy_unfil[[3]]

tidy_otu <- tidy_data[[1]]
tidy_full <- tidy_data[[2]]
tidy_humans <- tidy_data[[3]]
```


### Richness analysis
Richness analysis was performed with the `phyloseq::estimate_richness` function. Beta diversity was performed with PERMANOVA using the `adonis` function in the `vegan` package. 

```{r richness, fig.align="center"}
# source("./R/process_data.R")

richness_data <- phyloseq::estimate_richness(full_OTU)
richness_data <- data_frame(metric = as.vector(sapply(colnames(richness_data), rep, nrow(richness_data))),
                            alpha_diversity = as.vector(as.matrix(richness_data)),
                            sample_name = as.vector(t(sapply(rownames(richness_data), rep, ncol(richness_data)))),
                            condition = as.vector(t(sapply(sample_data(full_OTU)[, 6], rep, ncol(richness_data)))),
                            replicate = as.vector(t(sapply(sample_data(full_OTU)[, 7], rep, ncol(richness_data)))),
                            day = as.vector(t(sapply(sample_data(full_OTU)[, 5], rep, ncol(richness_data)))),
                            cond_day = as.vector(t(sapply(sample_data(full_OTU)[, 8], rep, ncol(richness_data)))))



# Figure 3a: Diversity analysis 
strip_labels <- c("1" = "Day 1",
                  "2" = "Day 2",
                  "3" = "Day 3")

richness_data %>% 
  dplyr::filter(., metric == "Observed", !is.na(condition)) %>% 
  ggplot() + 
  geom_boxplot(aes(x = condition, y = alpha_diversity), outlier.size = 0) +
  geom_point(aes(x = condition, y = alpha_diversity, fill = condition), pch = 21, size = 4, color = "black") +
  scale_fill_manual(values = c("#006699", "#00CCFF"), breaks = c("aerobic", "anaerobic")) + 
  labs(x=NULL,y="Observed Alpha Diversity") + 
  facet_grid(~ day, labeller = as_labeller(strip_labels)) +
  theme_bw() + 
  theme(panel.grid = element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 18),
        axis.text.y = element_text(size = 18,color="black"),
        axis.text.x = element_text(size = 18,color="black", angle = 90, hjust = 1, vjust = 0.5),
        axis.title = element_text(size = 24,face="bold"),
        legend.position = "none")
```

### Beta diversity analysis
```{r beta_diversity, fig.align="center"}
# PERMANOVA for beta diversity
# test for beta-diversity (between groups diversity)
# global
# get data from phyloseq obj
# metadata <- as(sample_data(sasan_otu_final), "data.frame")
# metadata <- as(sample_data(full_OTU), "data.frame")
metadata <- data.frame(tidy_otu$samples)

counts <- as.matrix(tidy_otu$otu_table)
d <- vegan::vegdist(x = t(counts), method = "bray") # <-- distance between samples!
# d <- vegan::vegdist(x = counts, method = "bray")

permanova_stats <- vegan::adonis(formula = t(counts) ~ condition,
                          data = metadata,
                          method = "bray",
                          permutations = 1000)

# per day
x1 <- metadata[which(metadata$cond_day == "aerobic1" | metadata$cond_day == "anaerobic1"), ]
y1 <- counts[, which(colnames(counts) %in% x1$filename)]
permanova_day1 <- vegan::adonis(formula = t(y1) ~ condition,
                         data = x1,
                         method = "bray",
                         permutations = 1000)

x1 <- metadata[which(metadata$cond_day == "aerobic2" | metadata$cond_day == "anaerobic2"), ]
y1 <- counts[, which(colnames(counts) %in% x1$filename)]
permanova_day2 <- vegan::adonis(formula = t(y1) ~ condition,
                         data = x1,
                         method = "bray",
                         permutations = 1000)

x1 <- metadata[which(metadata$cond_day == "aerobic3" | metadata$cond_day == "anaerobic3"), ]
y1 <- counts[, which(colnames(counts) %in% x1$filename)]
permanova_day3 <- vegan::adonis(formula = t(y1) ~ condition,
                         data = x1,
                         method = "bray",
                         permutations = 1000)

tmp <- data_frame(day = c(1, 2, 3),
                  p = c(permanova_day1$aov.tab$`Pr(>F)`[1],
  permanova_day2$aov.tab$`Pr(>F)`[1],
  permanova_day3$aov.tab$`Pr(>F)`[1]))

knitr::kable(tmp)
```

<!-- ### Generate data frames for additional analyses -->
```{r data_frames}
source("./R/generate_dfs.R")
```

### Comparison to human microbiome
```{r comp_human, fig.align="center"}
df3 %>% 
  tibble::add_column(., min_val = bstat[, 1], q1_val = bstat[, 2], mid_val = bstat[, 3], q2_val = bstat[, 4], max_val = bstat[, 5]) %>%
  dplyr::mutate(., counts = replace(counts, counts == 0, 1)) %>%
  dplyr::filter(., sample != "Hmb11", genus != "Unknown") %>%
  ggplot() + 
  geom_boxplot(aes(x = genus, y = log10(counts), fill = sample), outlier.size = 0) +
  geom_point(aes(x = genus, y = log10(counts), fill = sample), position = position_dodge(width = 0.75), pch = 21, color = "black", size = 2) +
  scale_color_manual(values = c("#006699", "#00CCFF", "#FF3300"),
                     breaks = c("Aerobic", "Anaerobic", "Human stool"),
                     labels = c("Aerobic", "Anaerobic", "Human stool"),
                     name = "Sample type") +
  scale_fill_manual(values = c("#006699", "#00CCFF", "#FF3300"),
                    breaks = c("Aerobic", "Anaerobic", "Human stool"),
                    labels = c("Aerobic", "Anaerobic", "Human stool"),
                    name = "Sample type") +
  scale_shape_manual(values = c(0, 5, 15),
                     breaks = c("Aerobic", "Anaerobic", "Human stool")) + 
  labs(x = NULL, y = "log10(total counts)") +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 20, face = "bold", color = "black"),
        panel.grid = element_blank())
```

### Changes in anaerobic chip across days
```{r anaerobic_days, fig.align="center"}
# Figure 4a: Changes in genera in the anaerobic chip across days
df2 %>% 
  dplyr::filter(., genus != "Unknown", sample_group == "anaerobic") %>% 
  ggplot() + 
  geom_boxplot(aes(x = genus, y = log10(counts), fill = day), position = "dodge", outlier.size = 0) +
  geom_point(aes(x = genus, y = log10(counts), fill = day), pch = 21, size = 2, position = position_dodge(width = 0.75)) +
  scale_fill_manual(values = c("#FFCC00", "#FF3300", "#996666"),
                     breaks = c(1, 2, 3),
                     labels = c("Day 1", "Day 2", "Day 3"),
                     name = NULL) +
  # scale_fill_manual(values = c("#006699", "#00CCFF", "#CCCCCC"), 
  #                   breaks = c("1", "2", "3"),
  #                   labels = c("Day 1", "Day 2", "Day 3"),
  #                   name = NULL) + 
  labs(y = "log10(counts)", x = NULL) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,size=18,color="black"),
        axis.text.y = element_text(size=18,color="black"),
        axis.title = element_text(size=20,face="bold",color="black"),
        panel.grid = element_blank(),
        strip.background = element_blank(),
        strip.text = element_text(size=12,color="black"))
```

### Comparison of chip microbiome at different O2 levels with liquid culture microbiome
```{r liquid_v_chip, fig.align="center"}
# Figure 4b: Comparison of chip (anaerobic or aerobic) to liquid culture
a1 <- which(df2$sample_group == "anaerobic" & df2$day == 3)
a2 <- which(df2$sample_group == "aerobic" & df2$day == 3)
ctr <- which(df2$sample_group == "Hmb11")

growth_df <- data_frame(genus = c(as.character(df2$genus[a1]),
                                  as.character(df2$genus[a2])),
                        diff_abundance = c(df2$counts[a1]/df2$counts[ctr],
                                           df2$counts[a2]/df2$counts[ctr]),
                        condition = c(rep("anaerobic",length(a1)),rep("aerobic",length(a2))))

growth_df$diff_abundance[growth_df$diff_abundance == 0] <- 1

growth_df %>% 
  dplyr::filter(., genus != "Unknown") %>%
  ggplot() + 
  geom_boxplot(aes(x = genus, y = log10(diff_abundance), fill = as.factor(condition)), outlier.size = 0) +
  geom_point(aes(x = genus, y = log10(diff_abundance), fill = as.factor(condition)), pch = 21, size = 2, position = position_dodge(width = 0.75)) +
  # geom_dotplot(aes(x = genus, y = log10(diff_abundance), fill = as.factor(condition)), stackdir = "center") +
  scale_fill_manual(values = c("#006699", "#00CCFF"), 
                    breaks = c("aerobic", "anaerobic"),
                    labels = c("Aerobic", "Anaerobic"),
                    name = NULL) + 
  labs(y = "log10(Differential abundance)", x = NULL) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,size=18,color="black"),
        axis.text.y = element_text(size=18,color="black"),
        axis.title = element_text(size=20,face="bold",color="black"),
        panel.grid = element_blank(),
        strip.background = element_blank(),
        strip.text = element_text(size=12,color="black"))
```

### Differential abundance in anaerobic vs aerobic conditions
```{r diff_abun, fig.align="center"}
source("./R/differential_abundance.R")

fold_data %>% 
  dplyr::filter(., genus != "Unknown") %>%
  ggplot() + 
  geom_point(aes(x = fct_reorder(genus, log2, .desc = TRUE), y = log2, color = as.factor(day)), size = 4) + 
  scale_color_manual(values = c("#FFCC00", "#FF3300", "#996666"),
                     breaks = c(1, 2, 3),
                     labels = c("Day 1", "Day 2", "Day 3"),
                     name = "Day") + 
  labs(x = NULL, y = "log2(fold difference)") + 
  theme_bw() + 
  theme(axis.text.x = element_text(size = 18, color = "black", hjust = 1, vjust = 0.5, angle = 90),
        axis.text.y = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 24, face = "bold", color = "black"),
        panel.grid = element_blank())
```


# Supplementary figures
### Relative abundances across samples
```{r fig_s6, fig.align="center"}
# Figure S6: Barplot for relative abundances across all samples
df2 %>%
  # tibble::add_column(., s_id = factor(df2$sample_group, levels = c("Hmb11", "anaerobic", "aerobic"))) %>% 
  tibble::add_column(., p_id = gsub("Hmb11 \\w* \\(replicate NA\\)", "Hmb11", gsub(" )", ")",paste(df2$sample_group,df2$day,"(replicate",df2$replicate,")")))) %>% 
  ggplot() + 
  geom_bar(aes(x = p_id, y = relative_abundance, fill = genus), stat = "identity", color = "black") + 
  scale_fill_manual(breaks = c("Unknown", "Sutterella", "Bilophila", "Akkermansia", "Blautia",
                               "Oscillospira", "Ruminococcus", "Enterococcus", "[Eubacterium]",
                               "Parabacteroides", "Bacteroides", "Citrobacter"),
                    values = c("#CCCCCC", "#FFCC00", "#99CCCC", "#33CC33", "#CC3399",
                               "#CC99CC", "#336699", "#66CCFF", "#669933",
                               "#FF6600", "#993300", "#CCCCFF"),
                    name = "Genus") + 
  labs(x = NULL, y = "Relative\nabundance") + 
  theme_bw() + 
  theme(axis.title = element_text(size = 20, face = "bold", color = "black"),
        axis.text.x = element_text(size = 12, angle = 90, color = "black", hjust = 1, vjust = 0.5),
        axis.text.y = element_text(size = 12, color = "black"),
        panel.grid = element_blank())
```

### Relative abundance in anaerobic chip across days
```{r fig_s7, fig.align="center"}
# Figure S7: Relative abundance on anaerobic chip across all days
# somewhat captured in figure 4a

df2 %>% 
  dplyr::filter(., sample_group == "anaerobic", genus != "Unknown") %>%
  # filter(., !is.na(condition)) %>%
  # dplyr::filter(., relative_abundance < 0.2) %>%
  ggplot() + 
  geom_boxplot(aes(x = genus, y = relative_abundance, fill = as.factor(day)), outlier.size = 0) +
  geom_point(aes(x = genus, y = relative_abundance, fill = as.factor(day)), pch = 21, size = 2, position = position_dodge(width = 0.75)) +
  # geom_dotplot(aes(x = genus, y = relative_abundance, fill = as.factor(day)), binaxis = "y", stackdir = "center", position = "dodge") +
  scale_fill_manual(breaks = c(1, 2, 3),
                    values = c("#FF9900", "#CC9933", "#996633"),
                    labels = c("Day 1", "Day 2", "Day 3"),
                    name = NULL) +
  # geom_point(aes(x = bacteria, y = relative_abundance, color = as.factor(day)), position = position_jitter(width = 0.2)) +
  # scale_color_manual(breaks = c(1, 2, 3),
  #                   values = c("#FF9900", "#CC9933", "#996633"),
  #                   labels = c("Day 1", "Day 2", "Day 3"),
  #                   name = "Day") +
  # facet_grid(. ~ condition, scales = "free") + 
  labs(y = "Relative abundance", x = NULL) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,size=12,color="black"),
        axis.text.y = element_text(size=12,color="black"),
        axis.title = element_text(size=20,face="bold",color="black"),
        panel.grid = element_blank(),
        strip.background = element_blank(),
        strip.text = element_text(size=12,color="black"))
```

### Heatmap representation comparing human stool vs gut-on-chip microbiomes
```{r heatmap_human_chip, fig.align="center"}
df3 %>% 
  dplyr::mutate(., counts = replace(counts, counts == 0, 1)) %>%
  dplyr::filter(., sample != "Hmb11", genus != "Unknown") %>%
  ggplot() + 
  geom_tile(aes(x = genus, y = sample, fill = log10(counts)), color = "black") +
  scale_fill_gradient(low = "#FFFFFF", high = "#006699", guide = "colorbar", na.value = "white", name = "log10(counts)") +
  labs(x = NULL, y = NULL) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 20, face = "bold", color = "black"),
        panel.grid = element_blank())

```